templates/components/multiagent/typescript/workflow/factory.ts

```diff
@@ -0,0 +1,152 @@
+    this.memory = new ChatMemoryBuffer({
+      llm: this.llm,
+      chatHistory: options.chatHistory,
+    });
+    this.tools = options?.tools ?? [];
+    this.systemPrompt = options.systemPrompt;
+    this.writeEvents = options?.writeEvents ?? true;
+    this.role = options?.role;
+    this.callbackManager = this.createCallbackManager();
+    this.stream = options.stream;
+
+    // add steps
+    this.addStep(StartEvent<AgentInput>, this.prepareChatHistory, {
+      outputs: InputEvent,
+    });
+    this.addStep(InputEvent, this.handleLLMInput, {
+      outputs: StopEvent,
+    });
+  }
+
+  private get chatHistory() {
+    return this.memory.getAllMessages();
+  }
+
+  private async prepareChatHistory(
+    ctx: Context,
+    ev: StartEvent<AgentInput>,
+  ): Promise<InputEvent> {
+    const { message, streaming } = ev.data.input;
+    ctx.set("streaming", streaming);
+    this.writeEvent(`Start to work on: ${message}`);
+    if (this.systemPrompt) {
+      this.memory.put({ role: "system", content: this.systemPrompt });
+    }
+    this.memory.put({ role: "user", content: message });
+    return new InputEvent({ input: this.chatHistory });
+  }
+
+  private async handleLLMInput(
+    ctx: Context,
+    ev: InputEvent,
+  ): Promise<StopEvent<string | ReadableStream<EngineResponse>>> {
+    const chatEngine = new OpenAIAgent({
+      tools: this.tools,
+      systemPrompt: this.systemPrompt,
+      chatHistory: this.chatHistory,
+    });
+
+    if (!ctx.get("streaming")) {
+      const response = await Settings.withCallbackManager(
+        this.callbackManager,
+        () => {
+          return chatEngine.chat({
+            message: ev.data.input.pop()!.content,
+          });
+        },
+      );
+      this.writeEvent("Finished task");
+      return new StopEvent({ result: response.message.content.toString() });
+    }
+
+    const response = await Settings.withCallbackManager(
+      this.callbackManager,
+      () => {
+        return chatEngine.chat({
+          message: ev.data.input.pop()!.content,
+          stream: true,
+        });
+      },
+    );
+    ctx.writeEventToStream({ data: new FunctionCallingStreamResult(response) });
+    return new StopEvent({ result: response });
+  }
+
+  private createCallbackManager() {
+    const callbackManager = new CallbackManager();
+    callbackManager.on("llm-tool-call", (event) => {
+      const { toolCall } = event.detail;
+      this.writeEvent(
+        `Calling tool "${toolCall.name}" with input: ${JSON.stringify(toolCall.input)}`,
+      );
+    });
+    callbackManager.on("llm-tool-result", (event) => {
+      const { toolCall, toolResult } = event.detail;
+      this.writeEvent(
+        `Getting result from tool "${toolCall.name}": \n${JSON.stringify(toolResult.output)}`,
+      );
+    });
+    return callbackManager;
+  }
+
+  private writeEvent(msg: string) {
+    if (!this.writeEvents) return;
+    this.stream.appendMessageAnnotation({
```

Comment:
don't directly write to the stream, instead send an event via the context, e.g.:

```
context.writeEventToStream(
    new MessageEvent({ msg }),
  );
```